### PR TITLE
LedgerEnclave in epbft

### LINKED ISSUE #458: LedgerReplay to use LedgerEnclave to append entries to ledger
Based on the discussion [here](https://github.com/microsoft/CCF/pull/449#discussion_r335045208) the below changes will take place:

- `LedgerWriter` will not register callbacks and a generic ctx pointer to write to the ledger. `Replica` will have the `LedgerEnclave` object passed to it during construction and it will in turn initialize `LedgerWriter` with the `LedgerEnclave` object. That way `LedgerWriter` becomes more concrete and we can re-use it as described next
- `LedgerReplay`, will use `LedgerWriter`'s `LedgerEnclave` to put entries into it's own ledger, same way as `raft` does now. The entry will be returned to `LedgerReplay` without the frame. Then we will go on to deserializing the entry into a pre-prepare and executing it (only pre-prepare for now)

### PR DESCRIPTION
**Pull request by olgavrou**
Resolves #458 

- `LedgerWriter` will not register callbacks and a generic ctx pointer to write to the ledger 
- Replica will have the `LedgerEnclave` object passed to it during construction and it will in turn initialize `LedgerWriter` with the `LedgerEnclave` object
- `LedgerReplay` will use `LedgerWriter`'s `LedgerEnclave` to put entries into it's own ledger, same way as Raft does now. The entry will be returned to `LedgerReplay` without the frame. Then we will go on to deserializing the entry into a pre-prepare and executing it
- `test_ledger_replay` doesn't need to write to files anymore
- removed ledger initialization from `replica_test` since we don't check the ledger in e2e tests now

### CODE REVIEW

**Comment by cimetrics** [comment]
images

File: ePBFT/src/pbft/libbyz/Replica.cpp [hunk 1]
```diff
@@ -304,8 +310,11 @@ bool Replica::apply_ledger_data(const std::vector<uint8_t>& data)
     else
     {
       LOG_DEBUG << "Received entries could not be processed. Received seqno: "
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/Replica.cpp]: Minor: clang-format on this line seems a bit off?

File: ePBFT/src/pbft/libbyz/keypair.h [hunk 2]
```diff
@@ -128,9 +128,9 @@ class KeyPair
    */
   std::vector<uint8_t> sign(unsigned char* d, size_t d_size) const
   {
-    uint8_t sig[max_sig_size];
-    sign(d, d_size, sig);
-    return {sig, sig + signature_size};
+    std::vector<uint8_t> sig(max_sig_size);
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/keypair.h]: Let's leave this as it was as `libbyz/keypair.h` and `tls/keypair.h` will have to be merged at some point.

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp [hunk 3]
```diff
@@ -7,13 +7,14 @@
...
 #include "Request.h"
+#include "consensus/ledgerenclave.h"
 #include "host/ledger.h"
 #include "network_mock.h"
 
 #include <cstdio>
 #include <doctest/doctest.h>
 
-static constexpr size_t TOTAL_REQUESTS = 1050;
+static constexpr size_t TOTAL_REQUESTS = 30;
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: Why this change?

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp [hunk 4]
```diff
@@ -67,11 +68,18 @@ NodeInfo get_node_info()
   return node_info;
 }
 
-void init_replica(std::vector<char>& service_mem)
+void create_replica(
+  std::vector<char>& service_mem,
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: `std::vector<uint8_t>`?

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp [hunk 5]
```diff
@@ -145,70 +150,120 @@ TEST_CASE("Test Ledger Replay")
...
+
+            initial_ledger.insert(
+              end(initial_ledger),
+              (uint8_t*)&s,
+              (uint8_t*)&s + sizeof(uint32_t));
+            initial_ledger.insert(end(initial_ledger), data, data + size);
+          }
+          break;
+          default:
+            REQUIRE(false);
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: We should also write an error message here.

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp [hunk 6]
```diff
@@ -145,70 +150,120 @@ TEST_CASE("Test Ledger Replay")
...
-        auto ledger = static_cast<asynchost::Ledger*>(ctx);
-        ledger->write_entry(data, size);
-      };
-
-    replica->register_append_ledger_entry_cb(
-      append_ledger_entry_cb, replay_ledger_io.get());
-
-    auto initial_ledger_io =
-      std::make_unique<asynchost::Ledger>(initial_ledger, wf);
+    std::vector<uint8_t> initial_ledger;
```
Comment (jumaffre) [inline ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: I find the interface for the `initial_ledger` a little weird. Is the intent is to have an in-memory ledger? In that case, should we have a common abstract ledger interface and implement an in-memory subclass? I might be reading this wrong though.

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp (hunk 3 above)
Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: Each entry has a size of 33094 and if we have 1050 requests that exceeds the ringbuffer circuit size. I could increase the circuit size, but this also makes the test quicker.

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp (hunk 4 above)
Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: Replica's constructor expects `char`. It could probably be made `uint8_t` but it is out of the scope of this PR.

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp (hunk 3 above)
Comment (jumaffre) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: Would you mind adding a comment for this?

Comment (jumaffre) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: Actually, perhaps `circuit_size_shift` (or just `circuit_size`) could depend on `TOTAL_REQUESTS`? And we keep `TOTAL_REQUESTS` small for the reasons you mentioned?

Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: lemme try that out :)

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp (hunk 6 above)
Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: You are reading it correctly, `initial_ledger` is used to simulate what the class `Ledger` in `ledger.h` does. I had split that class up in the previous PR with intentions of possibly abstracting it, as we had discussed. I can move in this direction of others agree to this.

Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: You are reading it correctly, `initial_ledger` is used to simulate what the class `Ledger` in `ledger.h` does. I had split that class up in the previous PR with intentions of possibly abstracting it, as we had discussed. I can move in this direction of others agree to this.

File: ePBFT/src/pbft/libbyz/keypair.h (hunk 2 above)
Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/keypair.h]: I had changed this because the sanitizer was complaining when in Debug mode, but I don't think it's needed actually. I have reverted it.

File: ePBFT/src/pbft/libbyz/Replica.cpp (hunk 1 above)
Comment (olgavrou) [reply ePBFT/src/pbft/libbyz/Replica.cpp]: The formatting check seems to pass so I don't know.

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp (hunk 6 above)
Comment (achamayou) [reply ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp]: I think this is a nice way to have an in-memory equivalent of what we otherwise have in the file, without complicating the code that goes into CCF. There's no use case for an in-memory implementation of the ledger, nor should the enclave code be able to tell where the entries are stored. I prefer to have this in test fixtures for that reason.
